Feat: conditional permissions behavior

yunginnanet opened the pull request:
## Overview

Current behavior asks for sudo password via stdin regardless on if we're root or not. 

This PR adds an `IsRoot` boolean to the configuration struct and adjusts the final message appropriately, as well as other changes listed below.

## Change Items
   - Added a check for whether the script is run as root and adjusted operations accordingly.
   - Enhanced error handling and user notifications for permissions and file operations.
   		- This involved writing a wrapper for `HikariKnight/ls-iommu/pkg/common.ErrorCheck` 
   - Implemented `ExecAndLogSudo` function to streamline running commands with elevated permissions.
   - Refactored logging execution flow for exec outputs.
   - Refactored `fileio.FileExist` to return both a boolean and an error.
   - Updated function comments to clarify their purpose and usage.
   - Ensured consistent commenting style and enhanced readability.
   - Added a test for `finalizeNotice` function to manually verify output messages.
   - Updated handling of old `debug.log` to rename to `debug_old.log` instead of deleting.

HikariKnight commented:
I also noticed that if we tell people that they can run quickpassthrough itself as root now, we will have to check if the `config` and `backup` folders are owned by root and tell them to re-run as root quickpassthrough is not run through sudo.

I originally designed it to be run as the normal user so i never actually tested running it as sudo then normal user 😅

yunginnanet commented:
Thanks for the feedback! I was actually just checking in to ping you to request comment, so I'm happy to see you replied! I'll take a look :)

yunginnanet commented on internal/pages/06_finalize.go:
```diff
@@ -151,10 +175,29 @@ func installPassthrough(config *configs.Config) {
 		fmt.Printf("%s\n", output)
 	}
 
+	execAndLogSudo := func(cmd string) {
```
take a look at my latest refactoring work, this should resolve this I believe

yunginnanet commented:
@HikariKnight take a look when you can, I believe I've addressed your comments

but this is a pretty heavy refactor so I definitely need a second pair of eyes

yunginnanet commented:
The behavior that actually lead to me starting on this was that I would actually receive silent errors when I ran this command as a user after running it as root. This last commit, https://github.com/HikariKnight/quickpassthrough/pull/28/commits/395f5ab6bca280e06031480270d50ad0afe942d8, finalizes my major changes for now I believe. 

Here's a demo of the behavior this commit replaces silent fatal errors with:

![Fatal error demonstration](https://tcp.ac/i/JMSUc.gif)

HikariKnight commented:
this gif looks nice, conveniently i am sick today so i surprisingly has freetime to look at the code once i have the paperwork done

HikariKnight commented:
This looks good, will do a test build locally later today and test in VMs to make sure things work properly in ubuntu, pop, arch and fedora since this is a big refactor.

yunginnanet commented:
> This looks good, will do a test build locally later today and test in VMs to make sure things work properly in ubuntu, pop, arch and fedora since this is a big refactor.

Yes, this definitely needs to be looked at carefully and tested. Thanks for your efforts, and take care of yourself.

yunginnanet commented:
^ The code should be easier to review now, as well.

yunginnanet commented on internal/pages/06_finalize.go:
```diff
@@ -145,68 +167,62 @@ func installPassthrough(config *configs.Config) {
+	if exists, _ := fileio.FileExist(modprobeFile); exists {
+		// Copy initramfs-tools module to system, note that CopyToSystem will log the command and output
+		// as well as check for errors
+		configs.CopyToSystem(config.IsRoot, modprobeFile, "/etc/modprobe.d/vfio.conf")
```
I didn't quite catch that we're doing it like that, I'll look into just copying it with golang if we're root

HikariKnight replied:
> I didn't quite catch that we're doing it like that, I'll look into just copying it with golang if we're root

well the config files are stored inside `config` which resides in the same directory as the binary.
saw no reason to make it plaster them in /etc since this is a tool you often would just run once and remove :)

yunginnanet replied:
Should work now! added a test case as well

yunginnanet replied:
and yeah as a side-note, some automation for testing would be amazing

may be something i can look into in the future

yunginnanet replied:
> i am happy i can at least test run the program inside a VM to catch this even though i dont use a mutable system anymore 😅

I honestly didn't think this would work properly in qemu for test purposes, if you have too many more failures I can just mark this as WIP again and fire up a VM on my end to run it though to spare you the time

HikariKnight replied:
> and yeah as a side-note, some automation for testing would be amazing
> 
> may be something i can look into in the future

would be awesome, i am still super green at golang, plus this isnt really the easiest project to make tests for since there is manual verification needed.

I am going to refactor the [ls-iommu](https://github.com/hikariknight/ls-iommu) project at a later date (hoping for end of summer if lucky) since it has a lot of ugly code that was mostly written to just have a functional version with some heavy nested ifs (made it just easier for my brain to visualize the logic as i wrote it), in the end that should give a nice performance bump too when that is done.

But its nice to have some extra eyes here on this project since this is the most front facing of the 2.

yunginnanet replied:
Yeah I spent some time looking through `ls-iommu` and actually started refactoring it a bit, but decided to reel in my ADHD brain back to this repo. I'm happy to help, I know how much of a daunting task it can be to go back and refactor older projects.

HikariKnight commented:
Tested and working on Fedora Workstation (Grubby and Dracut test)

HikariKnight commented on internal/configs/config_bootloaders.go:
```diff
@@ -70,39 +73,33 @@ func Set_Cmdline(gpu_IDs []string) {
-	return fmt.Sprintf("Executed: sudo kernelstub -a \"%s\"", kernel_args)
+	// Run and log, check for errors
+	common.ErrorCheck(command.ExecAndLogSudo(isRoot, true,
+		"kernelstub -a "+kernel_args,
```
`kernel_args` must be quoted,
![image](https://github.com/HikariKnight/quickpassthrough/assets/2557889/99cd8b67-5de8-4e8f-8591-e054b8ccb4d4)

so the output should have been
`sudo kernelstub -a "iommu=pt amd_iommu=on vfio_pci.ids=1af4:1050  vfio_pci.disable_vga=1"`

so we must change it to `'kernelstub -a "'+kernel_args+'"'`

HikariKnight commented:
Archlinux tested and works (pure grub and mkinitcpio test)

only need the pop os test to succeed then we should be good @yunginnanet

HikariKnight commented on internal/configs/config_bootloaders.go:
```diff
@@ -83,7 +83,7 @@ func Set_KernelStub(isRoot bool) {
 	// Run and log, check for errors
 	common.ErrorCheck(command.ExecAndLogSudo(isRoot, true,
-		"kernelstub -a "+kernel_args,
+		"kernelstub -a "+fmt.Sprintf("\"%s\"", kernel_args),
```
just out of curiosity, wouldn't `fmt.Sprintf("kernelstub -a \"%s\"", kernel_args),` be more readable, or is that just a preference thing?

anyways reinstalling my popos vm for testing.

HikariKnight replied:
The good news though, it seems like everything else afterwards will execute fine :tada:

HikariKnight commented on pkg/command/command.go:
```diff
@@ -71,29 +84,79 @@ func Elevate(password string) {
+		return err
+	}
+
+	cs := strings.Fields(cmd)
+	r := exec.Command(cs[0], cs[1:]...)
```
The good news though, it seems like everything else once we fix this will execute fine 🎉

yunginnanet commented:
agreed on all points. adapter function added, complete with unit test

HikariKnight commented on pkg/command/command.go:
```diff
@@ -71,29 +84,101 @@ func Elevate(password string) {
+	cmdBin := args[0]
+	args = args[1:]
+	for i, arg := range args {
+		if !strings.HasPrefix(arg, "\"") {
+			continue
+		}
+		// find the end of the quoted argument
+		for j, a := range args[i:] {
+			if strings.HasSuffix(a, "\"") {
+				args[i] = strings.Join(args[i:i+j+1], " ")
+				args = append(args[:i+1], args[i+j+1:]...)
+				break
```
After seeing this commit, i remembered i already fixed this in `command.Run`
a less complex way would be to just pass any parameters after cmd as args that get combined into a stringlist
```go
func ExecAndLogSudo(isRoot, noisy bool, binary string, args ...string) error {
```

this means we could use
```go
ExecAndLogSudo(isRoot, true, "kernelstub", "-a", kernel_args)
```
`kernelstub` will be the cmd, `-a` and the `kernel_args` string will get combined into a `[]string` that you can just expand the args when we call exec.Command `exec.Command(binary, args...)`, no processing needed.
See the old implemention in `func Run` in command.go :)

that is how the old implementation did it and it worked and didnt rely on processing the args and using quotes since every parameter after the cmd was an argument.

We could instead just make an if to see if we are root and then "promote" the `binary` variable to  an arg
`exec.Command("sudo", binary, args...)`
and if we are root just do
`exec.Command(binary, args...)`

i believe this would be much cleaner and i feel dumb for not even thinking about it on my last comment 🤦

HikariKnight replied:
actually you just did this with `RunErrSudo` so just need to do the same thing for `ExecAndLogSudo`

yunginnanet replied:
I don't disagree that there are better and simpler ways to do this, and this isn't the only area that statement rings true for. Out of curiosity, is it not working reliably as it sits?

HikariKnight replied:
> I don't disagree that there are better and simpler ways to do this, and this isn't the only area that statement rings true for. Out of curiosity, is it not working reliably as it sits?
> 
> In situations where a solution has unnecessary complexity I do my best to compensate with test coverage. I am not against refactoring it further but if I'm being honest if we were to get particular about simpler and more correct solutions we could be here all year 😅
> 
> That said, if it is currently still not working then I'll definitely jump back on it. (I will either way if you insist, of course)

Personally i think the simpler method just using stringlists would be the best and i am sorry you had to spend time on that other workaround.

I also feel it will be a lot safer than forcing in quotes and stripping them  (potentially creating unintended behavior) since every entry in a `[]string` is treated as if it is quoted if it contains spaces in terms of `exec.Command`.

I can merge as is and manually fix it myself as what you have already done is herculean for me and i really appreciate it! But i may not be able to write a test for it without me sitting down for a day and reading up properly on it (again).
Then i will just tag a new release once it is done. :smiley:

yunginnanet replied:
I certainly can't say I disagree with your points here. You don't need to be sorry, it was just the closest roll of proverbial duct tape I conjured up. I'll take a look into this

yunginnanet replied:
@HikariKnight 
latest changes should do it! added a much more realistic unit test to verify the behavior as well

HikariKnight replied:
thanks i will run a check later today if i have the time, i just got back from vacation :smile:

yunginnanet replied:
Awesome, no worries!

HikariKnight commented on internal/configs/configs.go:
```diff
@@ -223,29 +273,67 @@ func backupFile(source string) {
+	}
+
+	conffile = strings.ReplaceAll(conffile, " ", "\\ ")
+	cmd := fmt.Sprintf("cp -v %s %s", conffile, sysfile)
+
+	err := command.ExecAndLogSudo(isRoot, false, cmd)
```
I tested everything but i had to patch this into
```go
err := command.ExecAndLogSudo(isRoot, false, "cp", "-v", conffile, sysfile)
```

in order for everything to work as it got missed when we changed how ExecAndLogSudo worked. i will just patch that after merge
